test_launch_ros/test/test_launch_ros/test_normalize_parameters.py

```diff
@@ -0,0 +1,173 @@
+
+import pathlib
+
+from launch import LaunchContext
+from launch.substitutions import TextSubstitution
+from launch_ros.utilities import evaluate_parameters
+from launch_ros.utilities import normalize_parameters
+import pytest
+
+
+def test_not_a_list():
+    with pytest.raises(TypeError):
+        normalize_parameters({'foo': 'bar'})
+    with pytest.raises(TypeError):
+        normalize_parameters('foobar')
+
+
+def test_single_str_path():
+    norm = normalize_parameters(['/foo/bar'])
+    expected = (pathlib.Path('/foo/bar'),)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_single_pathlib_path():
+    norm = normalize_parameters([pathlib.Path('/foo/bar')])
+    expected = (pathlib.Path('/foo/bar'),)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_multiple_paths():
+    norm = normalize_parameters([pathlib.Path('/foo/bar'), '/bar/baz'])
+    expected = (pathlib.Path('/foo/bar'), pathlib.Path('/bar/baz'))
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_path_with_substitutions():
+    orig = [(TextSubstitution(text='/foo'), TextSubstitution(text='/bar'))]
+    norm = normalize_parameters(orig)
+    expected = (pathlib.Path('/foo/bar'),)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_single_dictionary():
+    orig = [{'foo': 1, 'bar': 2.0}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'bar': 2.0},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_multiple_dictionaries():
+    orig = [{'foo': 1, 'bar': 2.0}, {'baz': 'asdf'}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'bar': 2.0}, {'baz': 'asdf'})
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution():
+    orig = [{TextSubstitution(text='bar'): TextSubstitution(text='baz')}]
+    norm = normalize_parameters(orig)
+    expected = ({'bar': 'baz'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution_list_name():
+    orig = [{(TextSubstitution(text='bar'), TextSubstitution(text='foo')): 1}]
+    norm = normalize_parameters(orig)
+    expected = ({'barfoo': 1},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution_list_value():
+    orig = [{'foo': [TextSubstitution(text='fiz'), TextSubstitution(text='buz')]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'fizbuz'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+    orig = [{'foo': [[TextSubstitution(text='fiz')], [TextSubstitution(text='buz')]]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': ('fiz', 'buz')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_mixed_substitutions_and_strings():
+    orig = [{'foo': [TextSubstitution(text='fiz'), 'bar']}]
```

Comment:
We should add a test case with a string as the first element followed by a substitution.

Reply:
Added in 46b149528115662912592690150d5af577226867